include/ignition/gazebo/detail/View.hh

```diff
@@ -34,110 +36,250 @@ namespace gazebo
 inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
 namespace detail
 {
-/// \brief A key into the map of views
-using ComponentTypeKey = std::set<ComponentTypeId>;
-
-/// \brief A view is a cache to entities, and their components, that
-/// match a set of component types. A cache is used because systems will
-/// frequently, potentially every iteration, query the
-/// EntityComponentManager for sets of entities that match a set of
-/// component types. Rather than look up the entities every time, we can
-/// use a cache to improve performance. The assumption is that entities
-/// and the types of components assigned to entities change infrequently
-/// compared to the frequency of queries performed by systems.
-///
-/// Note that symbols for this class are visible because methods from this class
-/// are used in templated Ignition::Gazebo::EntityComponentManager methods.
-/// However, users should not use this class (or anything else in namespace
-/// ignition::gazebo::detail) directly.
-class IGNITION_GAZEBO_VISIBLE View
+/// \brief A view that caches a particular set of component type data.
+/// \tparam ComponentTypeTs The component type(s) that are stored in this view.
+template<typename ...ComponentTypeTs>
+class IGNITION_GAZEBO_VISIBLE View : public BaseView
 {
-  /// Get a pointer to a component for an entity based on a component type.
-  /// \param[in] _entity The entity.
-  /// \param[in] _ecm Pointer to the entity component manager.
-  /// \return Pointer to the component.
-  public: template<typename ComponentTypeT>
-          const ComponentTypeT *Component(const Entity _entity,
-              const EntityComponentManager *_ecm) const
+  /// \brief Alias for containers that hold and entity and its component data.
+  /// The component types held in this container match the component types that
+  /// were specified when creating the view.
+  private: using ComponentData = std::tuple<Entity, ComponentTypeTs*...>;
+  private: using ConstComponentData =
+           std::tuple<Entity, const ComponentTypeTs*...>;
+
+  /// \brief Constructor
+  public: View()
+  {
+    this->componentTypes = {ComponentTypeTs::typeId...};
+  }
+
+  /// \brief Documentation inherited
+  public: bool HasCachedComponentData(const Entity _entity) const override
```

Comment:
Nit: can we separate the function declarations from the implementations? See the old `detail/ComponentStorageBase.hh` file for an example. That makes it easier to see all the API.

Reply:
Yep :+1: see https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/ea7d9e9b281e30d03ca89d3b22879f56fe53194b